### PR TITLE
Fix(*) reject conflicting bootstrap when AdminPort is set

### PR DESCRIPTION
**Pull request by nickolaev**
### Summary

When `kuma-dp` is registered to the `kuma-cp` there may be a conflict between the AdminPort selected and one fo the listeners configured. Add an extra check for such a scenario and make sure the error propagates properly.

### Full changelog

* chore(*) error propagation for kuma-cp/kuma-dp communication
* fix(kuma-cp) verify the admin port does not conflict with the listeners
* chore(makefile) allow to use run/kuma-dp with dataplane tokens

### Issues resolved

None

### CODE REVIEW

File: pkg/xds/bootstrap/generator.go [hunk 1]
```diff
@@ -101,6 +106,22 @@ func (b *bootstrapGenerator) generateFor(proxyId core_xds.ProxyId, dataplane *co
 	return b.configForParameters(params)
 }
 
+func (b *bootstrapGenerator) verifyAdminPort(adminPort uint32, dataplane *core_mesh.DataplaneResource) error {
+	for _, inbound := range dataplane.Spec.Networking.Inbound {
+		if adminPort == inbound.Port {
+			return errors.Errorf("Resource precondition failed: Port %d requested as both admin and inbound port.", adminPort)
+		}
+	}
```
Comment (jakubdyszkiewicz) [inline pkg/xds/bootstrap/generator.go:114]: inbound can have an address specified therefore even with the same port there won't be a clash.
For inbounds, if address is not specified, then the address from networking is taken.

File: pkg/xds/bootstrap/generator.go [hunk 2]
```diff
@@ -101,6 +106,22 @@ func (b *bootstrapGenerator) generateFor(proxyId core_xds.ProxyId, dataplane *co
...
+
+	for _, outbound := range dataplane.Spec.Networking.Outbound {
+		if adminPort == outbound.Port {
+			return errors.Errorf("Resource precondition failed: Port %d requested as both admin and outbound port.", adminPort)
+		}
```
Comment (jakubdyszkiewicz) [inline pkg/xds/bootstrap/generator.go:119]: same for outbound. It defaults to `127.0.0.1` but can have other address

File: app/kuma-dp/pkg/dataplane/envoy/remote_bootstrap.go [hunk 3]
```diff
@@ -48,10 +48,17 @@ func (b *remoteBootstrap) Generate(url string, cfg kuma_dp.Config) (proto.Messag
...
+	if resp.StatusCode != http.StatusOK {
+		if resp.StatusCode == http.StatusNotFound {
 			return nil, errors.New("Dataplane entity not found. If you are running on Universal please create a Dataplane entity on kuma-cp before starting kuma-dp. If you are running on Kubernetes, please check the kuma-cp logs to determine why the Dataplane entity could not be created by the automatic sidecar injection.")
 		}
+		if resp.StatusCode == http.StatusUnprocessableEntity {
+			bodyBytes, err := ioutil.ReadAll(resp.Body)
+			if err != nil {
+				return nil, errors.Errorf("Unable to read the response with status code: %d", resp.StatusCode)
+			}
+			return nil, errors.Errorf("Error: %s", string(bodyBytes))
```
Comment (jakubdyszkiewicz) [inline app/kuma-dp/pkg/dataplane/envoy/remote_bootstrap.go:60]: Is `Error: ` necessary? What will be the output then when I execute kuma-dp and such error will happen?

File: app/kuma-dp/cmd/run.go [hunk 4]
```diff
@@ -120,7 +120,6 @@ func newRunCmd() *cobra.Command {
 			runLog.Info("starting Kuma DP", "version", kuma_version.Build.Version)
 			if err := componentMgr.Start(core.SetupSignalHandler()); err != nil {
 				runLog.Error(err, "error while running Kuma DP")
-				return err
```
Comment (jakubdyszkiewicz) [inline app/kuma-dp/cmd/run.go:123]: Why we are not exiting here anymore?

File: pkg/xds/bootstrap/generator.go [hunk 5]
```diff
@@ -101,6 +106,22 @@ func (b *bootstrapGenerator) generateFor(proxyId core_xds.ProxyId, dataplane *co
...
+	}
+
+	return nil
+}
+
```
Comment (jakubdyszkiewicz) [inline pkg/xds/bootstrap/generator.go:124]: Can we have some tests for it?

File: app/kuma-dp/cmd/run.go (hunk 4 above)
Comment (nickolaev) [reply app/kuma-dp/cmd/run.go:123]: Yes, let me revert that.

File: app/kuma-dp/pkg/dataplane/envoy/remote_bootstrap.go (hunk 3 above)
Comment (nickolaev) [reply app/kuma-dp/pkg/dataplane/envoy/remote_bootstrap.go:60]: The output is too convoluted in general. I wanted to make it slightly better, but that can mess with other errors.
`kuma-dp` just outputs that it crashed with too many Go internal details (backtrace dumps).

File: pkg/xds/bootstrap/generator.go (hunk 1 above)
Comment (nickolaev) [reply pkg/xds/bootstrap/generator.go:114]: Got it, thanks for the clarification. Fixing this.

File: pkg/xds/bootstrap/generator.go (hunk 2 above)
Comment (nickolaev) [reply pkg/xds/bootstrap/generator.go:119]: I see, will fix.

File: pkg/xds/bootstrap/generator.go (hunk 5 above)
Comment (nickolaev) [reply pkg/xds/bootstrap/generator.go:132]: Tests added.

File: pkg/xds/bootstrap/generator_test.go [hunk 6]
```diff
@@ -218,4 +218,121 @@ var _ = Describe("bootstrapGenerator", func() {
...
+		request = types.BootstrapRequest{
+			Mesh:      "mesh",
+			Name:      "name-2.namespace",
+			AdminPort: 9901,
+		}
+
+		// when
+		_, err = generator.Generate(context.Background(), request)
+		// then
+		Expect(err).To(HaveOccurred())
```
Comment (jakubdyszkiewicz) [inline pkg/xds/bootstrap/generator_test.go:286]: Can we check the error?

File: pkg/xds/bootstrap/generator_test.go [hunk 7]
```diff
@@ -218,4 +218,121 @@ var _ = Describe("bootstrapGenerator", func() {
...
+							Port:        443,
+							ServicePort: 8443,
+							Tags: map[string]string{
+								"service": "backend",
+							},
+						},
+					},
+					Outbound: []*mesh_proto.Dataplane_Networking_Outbound{
+						{
+							Address: "127.0.1.1",
```
Comment (jakubdyszkiewicz) [inline pkg/xds/bootstrap/generator_test.go:307]: I think I have some blinds spots about networking. Why `127.0.1.1:9901` and `127.0.0.1:9901` is a conflict?

Comment (nickolaev) [reply pkg/xds/bootstrap/generator_test.go:307]: Right, I was reading  RFC 5735, Section 3.
```
   127.0.0.0/8 - This block is assigned for use as the Internet host
   loopback address.  A datagram sent by a higher-level protocol to an
   address anywhere within this block loops back inside the host.
```
And kind of thought that in the end there will be a conflict. However, at least the Linux TCP/IP stack distinguishes between he IP addresses and allows for binding to different IP's in that range using the same port. Therefore - there is no conflict in this situation. Let me fix that.

File: pkg/xds/bootstrap/generator_test.go (hunk 6 above)
Comment (nickolaev) [reply pkg/xds/bootstrap/generator_test.go]: Yep, I did for all the errors here.

**Comment by nickolaev** [comment]
@jakubdyszkiewicz we can, and I did it. It actually revealed a wrong assumption in my logic. I think it is ready to merge now. Please review.

**Pull request merged**